### PR TITLE
AO3-4676 AO3-5903 Extend invitations tests and fix bug with invitations with blank emails

### PR DESCRIPTION
**Pull request by hatal175**
## Issue

https://otwarchive.atlassian.net/browse/AO3-4676
https://otwarchive.atlassian.net/browse/AO3-5903

## Purpose

This commit should extend coverage of invitations_controller to full. It also fixes a problem I ran into where a user could update his invitation to have an empty email and thus it goes into oblivion - the email is not sent but it does contain an empty string so it is not considered unsent.

I would like to request a review for invitation.rb to see it is a somewhat logical change in a system-side view. 

The invitation feature tests  have been moved to their own group, so the test pipeline on Codeship (and the page for it on the wiki) will need to be updated when this is merged.

```
# The expectation is that this script is copied in to the test area
# on Codeship rather than run as a single script as having timings for
# each stage is useful.
#
# Adding "|| :" to the end of a command will force it to succeed,
# useful for ignoring failed tests
export RAILS_ENV=test
export CUCUMBER_FORMAT=Ao3Cucumber::Formatter
./script/codeship/test_prepare.sh
./script/codeship/try.sh rspec "bundle exec rspec spec/controllers"
./script/codeship/try.sh rspec "bundle exec rspec spec/models"
./script/codeship/try.sh rspec "bundle exec rspec --exclude-pattern 'spec/{controllers,models}/**/*.rb'"
./script/codeship/test_prepare.sh
./script/codeship/try.sh admins             "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh bookmarks          "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh collections        "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh comments_and_kudos "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh gift_exchanges     "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh importing          "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh invitations        "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh other_a            "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh other_b            "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh prompt_memes_a     "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh prompt_memes_b     "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh prompt_memes_c     "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh search             "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh tag_sets           "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh tags_and_wrangling "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh users              "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh works              "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
```

## Testing

Run normal tests and see coverage.

## Credit

Tal Hayon

Please use he.

### CODE REVIEW

**Comment by sarken** [comment]
It looks like this has acquired some merge conflicts that will need to be resolved before it can be reviewed and merged.

File: spec/controllers/invitations_controller_spec.rb [hunk 1]
```diff
@@ -224,9 +226,24 @@
...
       delete :destroy, params: { id: invitation.id }
       it_redirects_to_with_notice(admin_invitations_path, "Invitation successfully destroyed")
     end
+
+    it "errors on failure" do
+      invite = create(:invitation, invitee_email: "")
+      admin.update(roles: ["policy_and_abuse"])
+      fake_login_admin(admin)
+
+      allow(Invitation).to receive(:find).with(invite.id.to_s).and_return(invite)
```
Comment (redsummernight) [inline spec/controllers/invitations_controller_spec.rb:239]: These two stubs can replace with `allow_any_instance_of(Invitation).to receive(:save).and_return(false)`.

File: features/invitations/invite_request.feature [hunk 2]
```diff
@@ -87,14 +87,51 @@ Feature: Invite requests
       And I should see how long I have to activate my account
       And I should see "If you haven't received this email within 24 hours"
 
+  Scenario: A user can't send an invite with an empty email
```
Comment (redsummernight) [inline features/invitations/invite_request.feature:90]: We can keep one of the new scenarios to test the error message "Please enter an email address." The other scenarios are better as specs, though.

**Review by redsummernight — CHANGES_REQUESTED** [review]

File: spec/controllers/invitations_controller_spec.rb (hunk 1 above)
Comment (sarken) [reply spec/controllers/invitations_controller_spec.rb:239]: It seems this results in an error:

```
Failure/Error: expect(Invitation).to have_received(:find)
       #<Invitation(id: integer, creator_id: integer, invitee_email: string, token: string, sent_at: datetime, created_at: datetime, updated_at: datetime, used: boolean, invitee_id: integer, invitee_type: string, creator_type: string, redeemed_at: datetime, from_queue: boolean, external_author_id: integer) (class)> expected to have received find, but that object is not a spy or method has not been stubbed.
```

Comment (sarken) [reply spec/controllers/invitations_controller_spec.rb:239]: Oh, wait, never mind. Read the whole test, self.

File: features/admins/admin_invitations.feature [hunk 3]
```diff
@@ -350,6 +350,18 @@ Feature: Admin Actions to Manage Invitations
       And I press "Update Invitation"
     Then I should see "oldman@ds9.com" in the "invitation_invitee_email" input
 
+  Scenario: An admin can delete an invitation
+    Given an invitation for "test@example.com" exists
+      And an invitation for "test@example.com" exists
```
[the remark this answers is not in the document]
Comment (sarken) [reply features/admins/admin_invitations.feature:356]: I think it takes you to a different page (the invitation show page, which apparently doesn't have the delete link) if there's only one. I updated the scenario to explain a bit.

File: features/invitations/invite_request.feature (hunk 2 above)
Comment (sarken) [reply features/invitations/invite_request.feature:90]: The "A user can't update an invalid invite with an empty email" one might not translate well into a spec... it's like
- submit the form on users/name/invitations with invalid info, 
- the invitation show view will render as a result
- fill in the form on the rendered show view incorrectly and submit that as well

**Pull request closed without merging**